### PR TITLE
Arm test fixes !

### PR DESCRIPTION
**Pull request by khancyr**
disable some tests on SUB
correct disarm_delay test and activate it only on copter
correct some arm message
prepare for https://github.com/ArduPilot/ardupilot/pull/9393

### CODE REVIEW

File: Tools/autotest/common.py [hunk 1]
```diff
@@ -560,7 +560,7 @@ def arm_motors_with_rc_input(self, timeout=20):
                 self.set_output_to_trim(self.get_rudder_channel())
                 self.progress("Arm in %ss" % arm_delay)  # TODO check arming time
                 return True
-        self.progress("FAILED TO ARM WITH RADIO")
+        self.progress("Unable to ARM with radio")
```
Comment (peterbarker) [inline Tools/autotest/common.py]: The original message was closer to the truth.  "Unable to" means it wasn't possible; "FAILED TO" means you attempted to and it didn't work.  Subtle but important difference.

Same applies to similar changes here.

File: Tools/autotest/common.py [hunk 2]
```diff
@@ -604,26 +604,53 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
-        self.progress("FAILED TO DISARM WITH SWITCH")
+        self.progress("Unable to DISARMED with switch")
         return False
 
-    def autodisarm_motors(self):
-        """Autodisarm motors."""
-        self.progress("Autodisarming motors")
-        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_GROUND_ROVER:  # NOT IMPLEMENTED ON ROVER
+    def wait_autodisarm_motors(self):
+        """Wait Autodisarm motors."""
```
Comment (peterbarker) [inline Tools/autotest/common.py]: This comment needs some help :-)

Also, this is most certainly not waiting for the vehicle to auto-disarm; it is actively disarming it, so it is now misnamed / misfeatured.

Comment (peterbarker) [reply Tools/autotest/common.py]: Better fix  may be to override in a subclass for any autodisarm test.

File: Tools/autotest/common.py [hunk 3]
```diff
@@ -604,26 +604,53 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
             self.progress("MOTORS AUTODISARMED OK")
             return True
+
+        if self.mav.mav_type in [mavutil.mavlink.MAV_TYPE_QUADROTOR,
+                                 mavutil.mavlink.MAV_TYPE_HELICOPTER,
+                                 mavutil.mavlink.MAV_TYPE_HEXAROTOR,
+                                 mavutil.mavlink.MAV_TYPE_OCTOROTOR,
+                                 mavutil.mavlink.MAV_TYPE_COAXIAL,
+                                 mavutil.mavlink.MAV_TYPE_TRICOPTER]:
+            disarm_delay = self.get_parameter("DISARM_DELAY", retry=2)
```
Comment (peterbarker) [inline Tools/autotest/common.py]: I can't see a good reason for specifying a retry value here.

File: Tools/autotest/common.py [hunk 4]
```diff
@@ -604,26 +604,53 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
             if not self.mav.motors_armed():
-                disarm_delay = self.get_sim_time() - tstart
+                disarm_time = self.get_sim_time() - tstart
                 self.progress("MOTORS AUTODISARMED")
-                self.progress("Autodisarm in %ss" % disarm_delay)  # TODO check disarming time
-                return True
-        self.progress("FAILED TO AUTODISARM")
-        return False
+                self.progress("Autodisarm in %ss, expect less than %ss" % (disarm_time, disarm_delay))
+                if disarm_time <= disarm_delay:
```
Comment (peterbarker) [inline Tools/autotest/common.py]: just return the result of the boolean expression.

File: Tools/autotest/common.py [hunk 5]
```diff
@@ -604,26 +604,53 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
+                    return True
+                else:
+                    return False
+        self.progress("Unable to AUTODISARMED")
```
Comment (peterbarker) [inline Tools/autotest/common.py]: Tense problem here; `Vehicle did not automatically disarm itself`, perhaps?

File: Tools/autotest/common.py [hunk 6]
```diff
@@ -604,26 +604,53 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
+        raise AutoTestTimeoutException()
+
+    def set_autodisarm_delay(self, delay):
```
Comment (peterbarker) [inline Tools/autotest/common.py]: You created this function and then didn't call it?

File: Tools/autotest/common.py [hunk 7]
```diff
@@ -719,6 +746,23 @@ def run_cmd(self,
                     raise ValueError()
                 break
 
+    def get_default_mode_string(self):
```
Comment (peterbarker) [inline Tools/autotest/common.py]: This looks like something the subclass might return instead.

File: Tools/autotest/common.py [hunk 8]
```diff
@@ -1279,6 +1323,8 @@ def init(self):
     def test_arm_feature(self):
         """Common feature to test."""
         self.context_push()
+        self.do_set_mode_via_command_long(self.get_default_mode_string())
```
Comment (peterbarker) [inline Tools/autotest/common.py]: `set_default_flightmode`

File: Tools/autotest/common.py [hunk 9]
```diff
@@ -1305,113 +1353,114 @@ def test_arm_feature(self):
...
+                    self.progress("Failed to NOT ARM")
+                    raise NotAchievedException()
+                if self.arm_motors_with_switch(arming_switch):
+                    self.progress("Failed to NOT ARM")
+                    raise NotAchievedException()
+                self.disarm_vehicle()
+                self.mav.wait_heartbeat()
+                self.set_rc(arming_switch, 1000)
+                self.set_rc(interlock_channel, 1000)
+                if self.mav.mav_type is mavutil.mavlink.MAV_TYPE_HELICOPTER:
```
Comment (peterbarker) [inline Tools/autotest/common.py]: I think I'd prefer factoring the test and overriding the factored component in the subclass.

File: Tools/autotest/common.py [hunk 10]
```diff
@@ -719,6 +719,23 @@ def run_cmd(self,
...
+                                 mavutil.mavlink.MAV_TYPE_HELICOPTER,
+                                 mavutil.mavlink.MAV_TYPE_HEXAROTOR,
+                                 mavutil.mavlink.MAV_TYPE_OCTOROTOR,
+                                 mavutil.mavlink.MAV_TYPE_COAXIAL,
+                                 mavutil.mavlink.MAV_TYPE_TRICOPTER]:
+            return "STABILIZE"
+        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_FIXED_WING:
+            return "MANUAL"
+        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_SUBMARINE:
+            return "STABILIZE"
```
Comment (jaxxzer) [inline Tools/autotest/common.py]: manual is the default mode for sub.

File: Tools/autotest/common.py (hunk 9 above)
Comment (jaxxzer) [reply Tools/autotest/common.py]: +1

File: Tools/autotest/common.py [hunk 11]
```diff
@@ -719,6 +746,23 @@ def run_cmd(self,
...
```
Comment (Williangalvani) [inline Tools/autotest/common.py]: I believe this should be MANUAL too,
With stabilize the sub autotest times out during one of the first wait_heading() calls

**Comment by khancyr** [comment]
rebase 
disable autodisarm test
adress most review
 close https://github.com/ArduPilot/ardupilot/issues/9682

File: Tools/autotest/common.py (hunk 1 above)
Comment (khancyr) [reply Tools/autotest/common.py]: arm_vehicle method is also using "unable". I don't want to use FAILED as we got test that need this to fail explicitly... Looking at failure on travis log is as simple as ctrl+f FAIL . If we keep this message we will also see the one that are wanted ...
I am ok to change the wording for something than fail !

**Comment by peterbarker** [comment]
@khancyr, sorry, I think I made this one conflict again.

**Comment by khancyr** [comment]
@peterbarker no problem ! I rebase and correct your mistake ;-P

File: Tools/autotest/common.py [hunk 12]
```diff
@@ -819,7 +819,7 @@ def disarm_motors_with_rc_input(self, timeout=20):
                 self.set_output_to_trim(self.get_rudder_channel())
                 self.progress("Disarm in %ss" % disarm_delay)  # TODO check disarming time
                 return True
-        self.progress("FAILED TO DISARM WITH RADIO")
+        self.progress("Unable to DISARMED with radio")
```
Comment (amilcarlucas) [inline Tools/autotest/common.py]: s/DISARMED/DISARM/

File: Tools/autotest/common.py [hunk 13]
```diff
@@ -846,26 +846,46 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
             if not self.mav.motors_armed():
                 self.progress("MOTORS DISARMED OK WITH SWITCH")
                 return True
-        self.progress("FAILED TO DISARM WITH SWITCH")
+        self.progress("Unable to DISARMED with switch")
```
Comment (amilcarlucas) [inline Tools/autotest/common.py]: s/DISARMED/DISARM/

File: Tools/autotest/common.py [hunk 14]
```diff
@@ -846,26 +846,46 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
-        self.progress("Autodisarming motors")
-        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_GROUND_ROVER:  # NOT IMPLEMENTED ON ROVER
-            self.progress("MOTORS AUTODISARMED OK")
-            return True
+    def wait_autodisarm_motors(self):
+        """Wait for Autodisarm motors within disarm delay
+        this feature is only available in copter (DISARM_DELAY) and plane (LAND_DISARMDELAY)."""
+        self.progress("Wait autodisarming motors")
+        disarm_delay = None
+        if self.mav.mav_type in [mavutil.mavlink.MAV_TYPE_QUADROTOR,
```
Comment (amilcarlucas) [inline Tools/autotest/common.py]: The DOCAHEXA is missing

File: Tools/autotest/common.py [hunk 15]
```diff
@@ -846,26 +846,46 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
-        return False
+                self.progress("Autodisarm in %ss, expect less than %ss" % (disarm_time, disarm_delay))
+                return disarm_time <= disarm_delay
+        self.progress("Unable to AUTODISARMED")
+        raise AutoTestTimeoutException()
+
+    def set_autodisarm_delay(self, delay):
+        """Set autodisarm delay"""
+        if self.mav.mav_type in [mavutil.mavlink.MAV_TYPE_QUADROTOR,
+                                 mavutil.mavlink.MAV_TYPE_HELICOPTER,
```
Comment (amilcarlucas) [inline Tools/autotest/common.py]: Again at least one copter frame is missing here.

File: Tools/autotest/common.py (hunk 14 above)
Comment (khancyr) [reply Tools/autotest/common.py]: not supported by mav_type

Comment (amilcarlucas) [reply Tools/autotest/common.py]: I think I have a PR for mavlink for that :)
@tridge can you merge it ?

**Comment by peterbarker** [comment]
I think the missing frame types thing shows we really shouldn't be enumerating them.

Happy to merge this as-is (assuming it passes!), but we should change to using callbacks onto the subclass to get information like the *name* of any disarm-delay parameter, that sort of thing.

**Comment by khancyr** [comment]
@peterbarker update with vehicle abstraction !

File: Tools/autotest/arducopter.py [hunk 16]
```diff
@@ -125,6 +125,12 @@ def close(self):
         if self.copy_tlog:
             shutil.copy(self.logfile, self.buildlog)
 
+    def get_rudder_channel(self):
+        return int(self.get_parameter("RCMAP_YAW"))
+
+    def get_disarm_delay(self):
+        return self.get_parameter("DISARM_DELAY")
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]: Technically this is an int as well.  Casting the results in the previous  getters is a bit gratuitous, 'though.

File: Tools/autotest/common.py [hunk 17]
```diff
@@ -680,7 +680,7 @@ def set_rc(self, chan, pwm, timeout=2000):
 
     def set_throttle_zero(self):
         """Set throttle to zero."""
-        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_GROUND_ROVER:
+        if self.is_rover():
```
Comment (peterbarker) [inline Tools/autotest/common.py]: Maybe `self.throttle_zero_value()` instead

File: Tools/autotest/common.py (hunk 1 above)
Comment (peterbarker) [reply Tools/autotest/common.py]: Use mixed-case `Failed to disarm with radio`

File: Tools/autotest/common.py [hunk 18]
```diff
@@ -846,26 +839,32 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
-                disarm_delay = self.get_sim_time() - tstart
+                disarm_time = self.get_sim_time() - tstart
                 self.progress("MOTORS AUTODISARMED")
-                self.progress("Autodisarm in %ss" % disarm_delay)  # TODO check disarming time
-                return True
-        self.progress("FAILED TO AUTODISARM")
-        return False
+                self.progress("Autodisarm in %ss, expect less than %ss" % (disarm_time, disarm_delay))
+                return disarm_time <= disarm_delay
+        self.progress("Unable to AUTODISARM")
```
Comment (peterbarker) [inline Tools/autotest/common.py]: Pass the string into the exception.

File: Tools/autotest/common.py [hunk 19]
```diff
@@ -846,26 +839,32 @@ def disarm_motors_with_switch(self, switch_chan, timeout=20):
...
+        raise AutoTestTimeoutException()
+
+    def set_autodisarm_delay(self, delay):
+        """Set autodisarm delay"""
```
Comment (peterbarker) [inline Tools/autotest/common.py]: Implement set_disarm_delay on subclasses, fatal in parent class to be consistent with getters.

File: Tools/autotest/common.py [hunk 20]
```diff
@@ -1096,20 +1095,15 @@ def mavproxy_do_set_mode_via_command_long(self, mode, timeout=30):
...
     def reach_heading_manual(self, heading):
         """Manually direct the vehicle to the target heading."""
-        if self.mav.mav_type in [mavutil.mavlink.MAV_TYPE_QUADROTOR,
-                                 mavutil.mavlink.MAV_TYPE_HELICOPTER,
-                                 mavutil.mavlink.MAV_TYPE_HEXAROTOR,
-                                 mavutil.mavlink.MAV_TYPE_OCTOROTOR,
-                                 mavutil.mavlink.MAV_TYPE_COAXIAL,
-                                 mavutil.mavlink.MAV_TYPE_TRICOPTER]:
+        if self.is_copter():
             self.mavproxy.send('rc 4 1580\n')
```
Comment (peterbarker) [inline Tools/autotest/common.py]: That's the rudder channel, right?

File: Tools/autotest/common.py [hunk 21]
```diff
@@ -1096,20 +1095,15 @@ def mavproxy_do_set_mode_via_command_long(self, mode, timeout=30):
...
             self.mavproxy.send('rc 4 1500\n')
             self.mav.recv_match(condition='RC_CHANNELS.chan4_raw==1500',
                                 blocking=True)
-        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_FIXED_WING:
+        if self.is_plane():
             self.progress("NOT IMPLEMENTED")
-        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_GROUND_ROVER:
+        if self.is_rover():
             self.mavproxy.send('rc 1 1700\n')
             self.mavproxy.send('rc 3 1550\n')
```
Comment (peterbarker) [inline Tools/autotest/common.py]: Should we really be putting throttle in?

File: Tools/autotest/common.py [hunk 22]
```diff
@@ -1122,20 +1116,15 @@ def reach_heading_manual(self, heading):
...
             self.mav.recv_match(condition='RC_CHANNELS.chan2_raw==1500',
                                 blocking=True)
-        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_FIXED_WING:
+        if self.is_plane():
             self.progress("NOT IMPLEMENTED")
-        if self.mav.mav_type == mavutil.mavlink.MAV_TYPE_GROUND_ROVER:
+        if self.is_rover():
             self.mavproxy.send('rc 3 1700\n')
             self.wait_distance(distance, accuracy=2)
             self.mavproxy.send('rc 3 1500\n')
```
Comment (peterbarker) [inline Tools/autotest/common.py]: Not really a fan of fiddling with the throttle values here.  The caller's just got no idea what's going on...

It's pre-existing, so don't fix it on this PR :-)

File: Tools/autotest/common.py [hunk 23]
```diff
@@ -1484,7 +1473,7 @@ def wait_text(self, text, timeout=20, the_function=None):
             if text.lower() in m.text.lower():
                 self.progress("Received expected text : %s" % m.text.lower())
                 return True
-        self.progress()
+        self.progress("")
```
Comment (peterbarker) [inline Tools/autotest/common.py]: Just remove this line?!

File: Tools/autotest/common.py [hunk 24]
```diff
@@ -1626,11 +1615,14 @@ def test_arm_feature(self):
         # TEST ARMING/DISARM
         self.set_parameter("ARMING_RUDDER", 2)  # allow arm and disarm with rudder on first tests
         interlock_channel = 8  # Plane got flighmode_ch on channel 8
-        if self.mav.mav_type is not mavutil.mavlink.MAV_TYPE_HELICOPTER:  # heli don't need interlock option
+        if not self.is_heli():  # heli don't need interlock option
```
Comment (peterbarker) [inline Tools/autotest/common.py]: s/don't need/needs/

File: Tools/autotest/common.py [hunk 25]
```diff
@@ -1677,31 +1657,26 @@ def test_arm_feature(self):
...
 
-        # Sub doesn't have 'stick commands'
-        if self.mav.mav_type != mavutil.mavlink.MAV_TYPE_SUBMARINE:
+            if self.is_copter():
+                self.start_test("Test arming failure with throttle too high")
+                self.set_rc(3, 1800)
+                try:
+                    if self.arm_vehicle():
+                        raise NotAchievedException("Armed when throttle too high")
+                except AutoTestTimeoutException():
```
Comment (peterbarker) [inline Tools/autotest/common.py]: I wonder if we can get rid of the `TimeoutException`?

File: Tools/autotest/common.py [hunk 26]
```diff
@@ -2002,81 +1971,26 @@ def test_gripper(self):
...
-    #         if m.alt <= 1 and home_distance < distance_min:
-    #             self.progress("RTL Complete")
-    #             return True
-    #     return False
-    #
-    # @abc.abstractmethod
-    # def test_mission(self, filename):
-    #     pass
+
+    def is_copter(self):
```
Comment (peterbarker) [inline Tools/autotest/common.py]: I was imagining that `arducopter.py` would override a base-class implementation which returns `False` to return `True`.

That way this list never gets out of date (which it is at the moment...)

**Comment by peterbarker** [comment]
This looks *so* much better with the `is_copter` stuff etc.

File: Tools/autotest/common.py (hunk 21 above)
Comment (khancyr) [reply Tools/autotest/common.py]: yes, that a normal rover so you need throttle to turn !

File: Tools/autotest/common.py (hunk 20 above)
Comment (khancyr) [reply Tools/autotest/common.py]: yep, wanna change that to set rudder ?

File: Tools/autotest/common.py (hunk 24 above)
Comment (khancyr) [reply Tools/autotest/common.py]: the comment is right, heli don't need to set explicitly interlock, as it is by default on channel 8

File: Tools/autotest/common.py (hunk 25 above)
Comment (khancyr) [reply Tools/autotest/common.py]: yep we can

**Comment by khancyr** [comment]
@peterbarker fix from  review

**Comment by peterbarker** [comment]
On Thu, 7 Feb 2019, Pierre Kancir wrote:

> yes, that a normal rover so you need throttle to turn !

I'm suggesting that the caller should ensure that movement if required. 
This Rover could be an omni or a skid-steerer.

**Comment by khancyr** [comment]
@peterbarker we don't have simple way to do that. I prefer to keep that for another PR, as it would need more change to handle all rover config (on spot turn, 2 paddle, omni etc)

File: Tools/autotest/arducopter.py [hunk 27]
```diff
@@ -1200,7 +1212,7 @@ def fly_auto_test(self):
         self.wait_waypoint(0, num_wp-1, timeout=500)
 
         # set throttle to minimum
-        self.set_rc(3, 1000)
+        self.set_throttle_zero_value()
```
Comment (peterbarker) [inline Tools/autotest/arducopter.py]: Maybe just `zero_throttle`

**Comment by khancyr** [comment]
@peterbarker changes done ! (change unable to failed and rename throttle_zero)

**Comment by peterbarker** [comment]
Squashed and merged, thanks!

**Pull request closed without merging**